src/ndarray.jl

```diff
@@ -396,13 +417,36 @@ function getindex(arr :: NDArray, ::Colon)
 end
 
 """
-Shortcut for [`slice`](@ref). **NOTE** the behavior for Julia's built-in index slicing is to create a
-copy of the sub-array, while here we simply call `slice`, which shares the underlying memory.
+Shortcut for [`slice`](@ref).
+**NOTE** the behavior for Julia's built-in index slicing is to create a
+copy of the sub-array, while here we simply call `slice`,
+which shares the underlying memory.
 """
 function getindex(arr :: NDArray, idx::UnitRange{Int})
   slice(arr, idx)
 end
 
+function getindex(arr :: NDArray, idx::Integer)
+  # linear indexing
+  len = length(arr)
+  size_ = size(arr)
+
+  if idx <= 0 || idx > len
+    throw(BoundsError(
+      "attempt to access $(join(size_, 'x')) NDArray at index $(idx)"))
+  end
+
+  idx -= 1
+  offsets = size_[1:end-1] |> reverse ∘ cumprod ∘ collect
+  handle = arr.handle
+  for offset ∈ offsets
+    handle = _at(handle, idx ÷ offset)
+    idx %= offset
+  end
+
+  _at(handle, idx) |> MX_NDArrayHandle |> x -> NDArray(x, arr.writable)
```

Comment:
It is kinda weird to return a one element array and also not common in Julia. The problem is if we convert it to a Julia value at this point we impose a synchronisation barrier.

Reply:
I agree with the weirdness, Python provide an API `nd.asscalar()` for conducting read value.
I think in Julia, we can have `first` for reading it.

looks like this
```julia
x[42] |> first  # do synchronisation and get the value
```

and keep `x[42]` returning an `NDArray`

Reply:
I can send a PR for `first` api. If you agree with that design

Reply:
+1 for the `first` design. I think we should make it explicit for users that an `NDArray` is not some transparent array, and it could potentially be on GPU and could potentially be slow to do indexing into every single element of it.

BTW: I feel I'm pretty out-dated about the latest Julian things when I see the `|>` and `∘` operators. :D

Reply:
`first` implemented @ 62bd213